AM5/ESM45 needed adjustments

ceblanton opened the pull request:
## Describe your changes
1. Use xarray for combining climatologies instead of `cdo merge`. `cdo merge` does not work for more than 100 variables, and even with fewer variables inserts duplicate coordinate variables. It's possible there are slight metadata issues but the resulting climos work fine for analysis scripts
2. `fre app mask-atmos-plevel` needs to trigger on pressure_mask = `.false.` instead of `False` and update test case. Also, keep the _unmsk legacy behavior trigger, but remove it from the error message so users don't see it.
3. Add static history files to the regrid task list

## Issue ticket number and link #677

## Checklist before requesting a review

- [x] I ran my code
- [x] I tried to make my code readable
- [x] I tried to comment my code
- [ ] I wrote a new test, if applicable
- [ ] I wrote new instructions/documentation, if applicable
- [ ] I ran pytest and inspected it's output
- [ ] I ran pylint and attempted to implement some of it's feedback
- [ ] No print statements; all user-facing info uses logging module

Commit:
Use xarray instead of CDO for merging the climatologies

Commit:
Use xarray instead of CDO for merging the climatologies

Commit:
Regrid the static history files as well as the temporal ones

Commit:
Regrid the static history files as well as the temporal ones

Commit:
Update catalogbuilder import structure after upstream change

Commit:
Set time to be a record dimension so analysis scripts can use them

Commit:
Set time to be a record dimension so analysis scripts can use them

Commit:
Modify fre pp mask-atmos-plevel to accept ".false." instead of "false" which is now users are writing it.
Also, remove the _unmsk note from the error message so folks don't use it

Commit:
Merge remote-tracking branch 'origin/am5-esm45-fixes' into am5-esm45-fixes

Commit:
Adjust mask-atmos-plevel test case
.false. is now replacing False as the variable attribute trigger

Commit:
Add static history file parsing to regrid-xy test

ceblanton commented on fre/tests/test_files/reduced_ascii_files/atmos_cmip.ua_unmsk.cdl:
```diff
@@ -50,7 +50,7 @@ variables:
 		ua_unmsk:standard_name = "eastward_wind" ;
 		ua_unmsk:interp_method = "conserve_order2" ;
-        ua_unmsk:pressure_mask = "False" ;
+        ua_unmsk:pressure_mask = ".false." ;
```
You caught me. `.false.` is the desired target exactly as that is that the diag manager prefers for booleans as configured, but looks goofy, and why not accept it without the periods you're saying.

```
  - module: tracers_plev19
    var_name: co2_vmr
    output_name: co2
    reduction: average
    kind: r4
    attributes:
    - pressure_mask: .false.
      cmip7_name: atmos.co2.tavg-p19-hxy-air.mon.glb
```

Commit:
After a second thought (thanks Ian) we will use "false" as the trigger not ".false."

Commit:
Revert "Adjust mask-atmos-plevel test case"

This reverts commit 0fe44fda6bb826a0939c7aa1f5ab77d0bc8e701d.

ilaflott commented on fre/app/regrid_xy/regrid_xy.py:
```diff
@@ -326,11 +326,18 @@ def regrid_xy(yamlfile: str,
+                    if static_history["source"] == source:
+                        components.append(component)
+            except KeyError:
+                pass
```
silent passing drives people (including me) nuts! maybe:

```suggestion
            except KeyError:
                fre_logger.warning( ... )
                pass
```

ceblanton replied:
I changed this one to a humble `if` also. I don't agree with the supposedly pythonic "Easier to Ask for Forgiveness than Permission" than "Look Before You Leap" either.

ilaflott commented on fre/app/generate_time_averages/combine.py:
```diff
@@ -39,21 +40,29 @@ def form_bronx_directory_name(frequency: str,
-    files = glob.glob(target)
-    if len(files) >= 1:
-        fre_logger.debug("%s has %s files", target, len(files))
+    input_files = glob.glob(source)
```
the term `source` is being used too often and too inconsistently. Here, `source` is really just a string input to `glob`, which uses it to find "source files".

Thus, if anything, `input_files` is actually the `source`'s, not the input to `glob`

ceblanton replied:
This method is much improved for input and output name clarity now, but there's plenty more where this came from, ha.

ilaflott commented on fre/app/mask_atmos_plevel/mask_atmos_plevel.py:
```diff
@@ -103,7 +103,7 @@ def mask_atmos_plevel_subtool(infile: str = None,
 
             fre_logger.info("Finished processing %s, wrote %s, pressure_mask is True", var, masked_var)
         else:
-            fre_logger.debug("Not processing %s, no pressure_mask attr, nor _unmsk in the variable name", var)
```
why would we hide an error message that clearly explains the code's behavior?

ceblanton replied:
Done, upgraded to `info`

ilaflott reviewed (changes requested):
i had more thoughts

ceblanton commented on fre/app/regrid_xy/tests/test_regrid_xy.py:
```diff
@@ -108,6 +121,19 @@ def test_regrid_xy():
+
+    assert np.all(test["mister"].values==np.float64(1.0))
+    assert np.all(test["darcy"].values==np.float64(2.0))
+    assert np.all(test["wins"].values==np.float64(3.0))
```
I'll agree the static test cases here do not give any additional functionality, and kind of confuse things. I would argue that static regrid-xy functionality can remain untested as it's a subset of temporary regrid-xy which is already tested.

ilaflott replied:
... can we not just make a second separate regrid static function?

ilaflott replied:
as a unit test, i mean

ceblanton commented on fre/app/regrid_xy/tests/generate_files.py:
```diff
@@ -152,6 +152,13 @@ def make_data():
+        for i in range(1, ntiles+1):
+          dataset.to_netcdf(f"{input_dir}/{date}.{history_file}.tile{i}.nc")
+    except KeyError:
+        pass
```
I changed it to a humble if. For the record, I agree with you on the silent pass, and I'll stop trying to be pythonic on this point.

Snippet from google, "why does python prefer try and except compared to if else"

Python generally prefers try and except (also known as EAFP - "Easier to Ask for Forgiveness than Permission") over extensive if and else (LBYL - "Look Before You Leap") for several reasons related to code readability, efficiency, and the language's core philosophy [1]. 

Python's Philosophy (EAFP)
The primary reason is the "Easier to Ask for Forgiveness than Permission" (EAFP) coding style, which is deeply ingrained in the language's design. This philosophy encourages attempting an operation first and handling the error if it fails, rather than pre-checking all possible error conditions beforehand [1].

ilaflott replied:
i like try/except- indeed, forgiveness wins with python, generally. i use it all the time!

testing is about rigor, though. look before you leap when testing, even in python. testing code shouldn't be written like normal code.

Commit:
Address reviewer feedback
- Improve merge-netcdf variable names for clarity
- Promote useful mask-atmos-plevel debug logging message to info
- Rephrase two try/except's as a humble if

Commit:
Revert "Add static history file parsing to regrid-xy test"

This reverts commit d4ab40ac4f932baeb79720b60e7040070690f45c.

Commit:
Add back a static regridding test case

Commit:
Merge remote-tracking branch 'origin/main' into am5-esm45-fixes

ceblanton commented:
@ilaflott I don't know why I had to change the assert line you touched to make it pass but it passes now.

ilaflott committed:
Add conditional skip for test_regrid_xy_static

Skip test if fregrid is not available in the environment.

ilaflott committed:
Merge branch 'main' into am5-esm45-fixes

ceblanton commented on fre/app/mask_atmos_plevel/mask_atmos_plevel.py:
```diff
@@ -80,9 +80,9 @@ def mask_atmos_plevel_subtool(infile: str = None,
                 ds_out[var] = mask_field_above_surface_pressure(ds_in, var, ds_ps)
                 ds_out[var].attrs['pressure_mask'] = "True"
-                fre_logger.info("Finished processing %s, pressure_mask is True", var)
+                fre_logger.info(f"Finished processing '{var}'. pressure_mask was False, now set to True")
```
I'll use the old-style prints in logging statements from now on :)
